frugen.c

```diff
@@ -460,6 +530,101 @@ int main(int argc, char *argv[])
 					fatal("JSON support was disabled at compile time");
 #endif
 				}
+				else if (use_binary) {
+					int fd = open(optarg, O_RDONLY);
+					if (fd < 0) {
+						perror("");
+						fatal("Failed to open file");
+					}
+
+					char common_header[8];
+					safe_read(fd, common_header, 8);
+
+					// Common Header Format Version = common_header[0]
+					// Internal Use Area Starting Offset = common_header[1]
+					uint16_t chassis_start_offset = 8 * common_header[2];
+					uint16_t board_start_offset = 8 * common_header[3];
+					uint16_t product_start_offset = 8 * common_header[4];
+					// MultiRecord Area Starting Offset = 8 * common_header[5]
+					// Padding = common_header[6] = 0
+					// Checksum = common_header[7]
+
+					// For the above offsets, an offset of 0 is valid and implies
+					// that the field is not present.
+
+					bool data_has_chassis = chassis_start_offset != 0;
+					bool data_has_board = board_start_offset != 0;
+					bool data_has_product = product_start_offset != 0;
+
+					if (data_has_chassis) {
```

Comment:
Please don't do this in `main()`. Implement functions to decode fru into `fru_exploded_*_t` structures, put them into `fru.c`, and use them here. Those must be the functions opposite to the existing `fru_chassis_info()`, `fru_product_info()` and `fru_board_info()`. I would propose to rename those to `fru_encode_*` and then add `fru_decode_*` ones. There also must be a counterpart to `fru_create()` that would disassemble the binary into separate areas.

Please keep in mind that the next step will be to add a true binary-to-json decoder. Make your changes as suitable for that task as possible.